docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,371 @@
+1. Thin-edge must be able to generate unique `id`s for child devices on the local network,
+   without consulting with the cloud.
+   The device registration service must work even if the tedge device is not connected to a cloud.
+   All translation between tedge-local-ids and cloud-twin-ids should be done internally by the cloud mapper.
+   It is okay for the mapper to use the same tedge-local-ids in the cloud, if they are unique enough.
+   But, the cloud IDs (typically very long and cumbersome) must not be exposed to the tedge network at any cost.
+1. A registration mechanism for thin-edge plugins to declare their capabilities and other necessary metadata to thin-edge.
+1. Enable a device to subscribe to all the data meant for itself, its services and all its descendent child devices.
+   This subscription mechanism should prevent a device from subscribing to messages of all other devices.
+
+### Nice-to have
+
+1. Avoid using the device id in the topics for the main tedge device to keep those well-known topics simple.
+   Use aliases like `main` or `master` instead, if an identifier is really required.
+   Using `id`s for child devices might be unavoidable.
+1. Consistency/symmetry in the topic structures for the main device, child device and services.
+   For e.g: if the telemetry topic for a child device is like: `tedge/level2/level3/level4/telemetry/...`
+   it would be better if the `telemetry` subtopic is at the 5th level in the topics for the main device and services as well,
+   as that would make wildcard subscriptions like "all telemetry data from all sources" easier.
+1. Dynamic creation/registration of child devices on receipt of the very first data that they send.
+   This is desired at least for immediate child devices, if not for further nested child devices.
+1. Easy to create static routing rules so that it is easy to map a local MQTT topic for a nested child device
+   to the equivalent cloud topic for its twin.
+1. It would be ideal if the context/source of data (tedge device, service or child device)
+   can be understood from the topic directly.
+   For e.g: a topic scheme like `tedge/main/{id}`, `tedge/service/{id}` and `tedge/child/{id}` is more user-friendly
+   than a simpler context agnostic scheme like `tedge/{id}` where `id` can be for any "thing".
+1. Limit the topic levels to 7 as AWS IoT core has a [max limit of 7](https://docs.aws.amazon.com/whitepapers/latest/designing-mqtt-topics-aws-iot-core/mqtt-design-best-practices.html)
+
+### Out of scope
+
+1. Routing different kinds of data to different clouds, e.g: all telemetry to azure and all commands from/to Cumulocity.
+   Even though this requirement is realistic, thin-edge MQTT topics must not be corrupted with cloud specific semantics,
+   and the same requirement will be handled with some external routing mechanism(e.g: routing via bridge topics)
+1. Make the `tedge/` MQTT topic prefix configurable so that it can be changed in the rare event of a name clash.
```

Comment:
I think we can drop this sentence all together. The `tedge/` topic prefix is a name-spacing mechanism which should be configurable in each component as it would allow running multiple thin-edge.io instances (and associated components) on a single MQTT broker. But the key point is that all components that need to interact with each other must use the name prefix.